# LAB | Feature Engineering

**Load the data**

In this challenge, we will be working with the same Spaceship Titanic data, like the previous Lab. The data can be found here:

https://raw.githubusercontent.com/data-bootcamp-v4/data/main/spaceship_titanic.csv

Metadata

https://github.com/data-bootcamp-v4/data/blob/main/spaceship_titanic.md

In [146]:
#Libraries
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsRegressor
from sklearn.preprocessing import MinMaxScaler, StandardScaler

In [147]:
spaceship = pd.read_csv("https://raw.githubusercontent.com/data-bootcamp-v4/data/main/spaceship_titanic.csv")
spaceship.head()

PassengerId HomePlanet CryoSleep  Cabin  Destination   Age    VIP  \
0     0001_01     Europa     False  B/0/P  TRAPPIST-1e  39.0  False   
1     0002_01      Earth     False  F/0/S  TRAPPIST-1e  24.0  False   
2     0003_01     Europa     False  A/0/S  TRAPPIST-1e  58.0   True   
3     0003_02     Europa     False  A/0/S  TRAPPIST-1e  33.0  False   
4     0004_01      Earth     False  F/1/S  TRAPPIST-1e  16.0  False   

   RoomService  FoodCourt  ShoppingMall     Spa  VRDeck               Name  \
0          0.0        0.0           0.0     0.0     0.0    Maham Ofracculy   
1        109.0        9.0          25.0   549.0    44.0       Juanna Vines   
2         43.0     3576.0           0.0  6715.0    49.0      Altark Susent   
3          0.0     1283.0         371.0  3329.0   193.0       Solam Susent   
4        303.0       70.0         151.0   565.0     2.0  Willy Santantines   

   Transported  
0        False  
1         True  
2        False  
3        False  
4         True

**Check the shape of your data**

In [148]:
spaceship.shape
spaceship.describe()

Age   RoomService     FoodCourt  ShoppingMall           Spa  \
count  8514.000000   8512.000000   8510.000000   8485.000000   8510.000000   
mean     28.827930    224.687617    458.077203    173.729169    311.138778   
std      14.489021    666.717663   1611.489240    604.696458   1136.705535   
min       0.000000      0.000000      0.000000      0.000000      0.000000   
25%      19.000000      0.000000      0.000000      0.000000      0.000000   
50%      27.000000      0.000000      0.000000      0.000000      0.000000   
75%      38.000000     47.000000     76.000000     27.000000     59.000000   
max      79.000000  14327.000000  29813.000000  23492.000000  22408.000000   

             VRDeck  
count   8505.000000  
mean     304.854791  
std     1145.717189  
min        0.000000  
25%        0.000000  
50%        0.000000  
75%       46.000000  
max    24133.000000

**Check for data types**

In [149]:
spaceship.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8693 entries, 0 to 8692
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   PassengerId   8693 non-null   object 
 1   HomePlanet    8492 non-null   object 
 2   CryoSleep     8476 non-null   object 
 3   Cabin         8494 non-null   object 
 4   Destination   8511 non-null   object 
 5   Age           8514 non-null   float64
 6   VIP           8490 non-null   object 
 7   RoomService   8512 non-null   float64
 8   FoodCourt     8510 non-null   float64
 9   ShoppingMall  8485 non-null   float64
 10  Spa           8510 non-null   float64
 11  VRDeck        8505 non-null   float64
 12  Name          8493 non-null   object 
 13  Transported   8693 non-null   bool   
dtypes: bool(1), float64(6), object(7)
memory usage: 891.5+ KB


Check every column in the dataset and review if it has the correct data type.

**Check for missing values**

In [150]:
spaceship.isna().sum()
#spaceship.isna()

PassengerId       0
HomePlanet      201
CryoSleep       217
Cabin           199
Destination     182
Age             179
VIP             203
RoomService     181
FoodCourt       183
ShoppingMall    208
Spa             183
VRDeck          188
Name            200
Transported       0
dtype: int64

There are multiple strategies to handle missing data

- Removing all rows or all columns containing missing data.
- Filling all missing values with a value (mean in continouos or mode in categorical for example).
- Filling all missing values with an algorithm.

For this exercise, because we have such low amount of null values, we will drop rows containing any missing value. 

In [151]:
spaceship_cleaned = spaceship.dropna(how="any")
spaceship_cleaned.info()

<class 'pandas.core.frame.DataFrame'>
Index: 6606 entries, 0 to 8692
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   PassengerId   6606 non-null   object 
 1   HomePlanet    6606 non-null   object 
 2   CryoSleep     6606 non-null   object 
 3   Cabin         6606 non-null   object 
 4   Destination   6606 non-null   object 
 5   Age           6606 non-null   float64
 6   VIP           6606 non-null   object 
 7   RoomService   6606 non-null   float64
 8   FoodCourt     6606 non-null   float64
 9   ShoppingMall  6606 non-null   float64
 10  Spa           6606 non-null   float64
 11  VRDeck        6606 non-null   float64
 12  Name          6606 non-null   object 
 13  Transported   6606 non-null   bool   
dtypes: bool(1), float64(6), object(7)
memory usage: 729.0+ KB


In [152]:
spaceship_cleaned.isna().sum()

PassengerId     0
HomePlanet      0
CryoSleep       0
Cabin           0
Destination     0
Age             0
VIP             0
RoomService     0
FoodCourt       0
ShoppingMall    0
Spa             0
VRDeck          0
Name            0
Transported     0
dtype: int64

In [153]:
spaceship_cleaned.nunique()

PassengerId     6606
HomePlanet         3
CryoSleep          2
Cabin           5305
Destination        3
Age               80
VIP                2
RoomService     1093
FoodCourt       1293
ShoppingMall     984
Spa             1143
VRDeck          1106
Name            6590
Transported        2
dtype: int64

In [154]:
spaceship_cleaned

PassengerId HomePlanet CryoSleep     Cabin    Destination   Age    VIP  \
0        0001_01     Europa     False     B/0/P    TRAPPIST-1e  39.0  False   
1        0002_01      Earth     False     F/0/S    TRAPPIST-1e  24.0  False   
2        0003_01     Europa     False     A/0/S    TRAPPIST-1e  58.0   True   
3        0003_02     Europa     False     A/0/S    TRAPPIST-1e  33.0  False   
4        0004_01      Earth     False     F/1/S    TRAPPIST-1e  16.0  False   
...          ...        ...       ...       ...            ...   ...    ...   
8688     9276_01     Europa     False    A/98/P    55 Cancri e  41.0   True   
8689     9278_01      Earth      True  G/1499/S  PSO J318.5-22  18.0  False   
8690     9279_01      Earth     False  G/1500/S    TRAPPIST-1e  26.0  False   
8691     9280_01     Europa     False   E/608/S    55 Cancri e  32.0  False   
8692     9280_02     Europa     False   E/608/S    TRAPPIST-1e  44.0  False   

      RoomService  FoodCourt  ShoppingMall     Spa  VRDeck               Name  \
0             0.0        0.0           0.0     0.0     0.0    Maham Ofracculy   
1           109.0        9.0          25.0   549.0    44.0       Juanna Vines   
2            43.0     3576.0           0.0  6715.0    49.0      Altark Susent   
3             0.0     1283.0         371.0  3329.0   193.0       Solam Susent   
4           303.0       70.0         151.0   565.0     2.0  Willy Santantines   
...           ...        ...           ...     ...     ...                ...   
8688          0.0     6819.0           0.0  1643.0    74.0  Gravior Noxnuther   
8689          0.0        0.0           0.0     0.0     0.0    Kurta Mondalley   
8690          0.0        0.0        1872.0     1.0     0.0       Fayey Connon   
8691          0.0     1049.0           0.0   353.0  3235.0   Celeon Hontichre   
8692        126.0     4688.0           0.0     0.0    12.0   Propsh Hontichre   

      Transported  
0           False  
1            True  
2           False  
3           False  
4            True  
...           ...  
8688        False  
8689        False  
8690         True  
8691        False  
8692         True  

[6606 rows x 14 columns]

In [155]:
spaceship_cleaned['Cabin'].head()

0    B/0/P
1    F/0/S
2    A/0/S
3    A/0/S
4    F/1/S
Name: Cabin, dtype: object

- **Cabin** is too granular - transform it in order to obtain {'A', 'B', 'C', 'D', 'E', 'F', 'G', 'T'}

In [156]:
#your code here
spaceship_cleaned["Cabin"] = spaceship_cleaned["Cabin"].astype(str) # Converting everything into a string
spaceship_cleaned["Cabin"] = spaceship_cleaned["Cabin"].str[0]

/var/folders/rx/4vz0nngn7_q84td2symm90y40000gn/T/ipykernel_7745/1275168790.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  spaceship_cleaned["Cabin"] = spaceship_cleaned["Cabin"].astype(str) # Converting everything into a string
/var/folders/rx/4vz0nngn7_q84td2symm90y40000gn/T/ipykernel_7745/1275168790.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  spaceship_cleaned["Cabin"] = spaceship_cleaned["Cabin"].str[0]


- Drop PassengerId and Name

In [157]:
#your code here
spaceship_cleaned

PassengerId HomePlanet CryoSleep Cabin    Destination   Age    VIP  \
0        0001_01     Europa     False     B    TRAPPIST-1e  39.0  False   
1        0002_01      Earth     False     F    TRAPPIST-1e  24.0  False   
2        0003_01     Europa     False     A    TRAPPIST-1e  58.0   True   
3        0003_02     Europa     False     A    TRAPPIST-1e  33.0  False   
4        0004_01      Earth     False     F    TRAPPIST-1e  16.0  False   
...          ...        ...       ...   ...            ...   ...    ...   
8688     9276_01     Europa     False     A    55 Cancri e  41.0   True   
8689     9278_01      Earth      True     G  PSO J318.5-22  18.0  False   
8690     9279_01      Earth     False     G    TRAPPIST-1e  26.0  False   
8691     9280_01     Europa     False     E    55 Cancri e  32.0  False   
8692     9280_02     Europa     False     E    TRAPPIST-1e  44.0  False   

      RoomService  FoodCourt  ShoppingMall     Spa  VRDeck               Name  \
0             0.0        0.0           0.0     0.0     0.0    Maham Ofracculy   
1           109.0        9.0          25.0   549.0    44.0       Juanna Vines   
2            43.0     3576.0           0.0  6715.0    49.0      Altark Susent   
3             0.0     1283.0         371.0  3329.0   193.0       Solam Susent   
4           303.0       70.0         151.0   565.0     2.0  Willy Santantines   
...           ...        ...           ...     ...     ...                ...   
8688          0.0     6819.0           0.0  1643.0    74.0  Gravior Noxnuther   
8689          0.0        0.0           0.0     0.0     0.0    Kurta Mondalley   
8690          0.0        0.0        1872.0     1.0     0.0       Fayey Connon   
8691          0.0     1049.0           0.0   353.0  3235.0   Celeon Hontichre   
8692        126.0     4688.0           0.0     0.0    12.0   Propsh Hontichre   

      Transported  
0           False  
1            True  
2           False  
3           False  
4            True  
...           ...  
8688        False  
8689        False  
8690         True  
8691        False  
8692         True  

[6606 rows x 14 columns]

In [158]:
spaceship_cleaned = spaceship_cleaned.drop(columns= ["PassengerId", "Name"])

In [159]:
spaceship_cleaned

HomePlanet CryoSleep Cabin    Destination   Age    VIP  RoomService  \
0        Europa     False     B    TRAPPIST-1e  39.0  False          0.0   
1         Earth     False     F    TRAPPIST-1e  24.0  False        109.0   
2        Europa     False     A    TRAPPIST-1e  58.0   True         43.0   
3        Europa     False     A    TRAPPIST-1e  33.0  False          0.0   
4         Earth     False     F    TRAPPIST-1e  16.0  False        303.0   
...         ...       ...   ...            ...   ...    ...          ...   
8688     Europa     False     A    55 Cancri e  41.0   True          0.0   
8689      Earth      True     G  PSO J318.5-22  18.0  False          0.0   
8690      Earth     False     G    TRAPPIST-1e  26.0  False          0.0   
8691     Europa     False     E    55 Cancri e  32.0  False          0.0   
8692     Europa     False     E    TRAPPIST-1e  44.0  False        126.0   

      FoodCourt  ShoppingMall     Spa  VRDeck  Transported  
0           0.0           0.0     0.0     0.0        False  
1           9.0          25.0   549.0    44.0         True  
2        3576.0           0.0  6715.0    49.0        False  
3        1283.0         371.0  3329.0   193.0        False  
4          70.0         151.0   565.0     2.0         True  
...         ...           ...     ...     ...          ...  
8688     6819.0           0.0  1643.0    74.0        False  
8689        0.0           0.0     0.0     0.0        False  
8690        0.0        1872.0     1.0     0.0         True  
8691     1049.0           0.0   353.0  3235.0        False  
8692     4688.0           0.0     0.0    12.0         True  

[6606 rows x 12 columns]

- For non-numerical columns, do dummies.

In [160]:
dummies = pd.get_dummies(spaceship_cleaned, columns=["HomePlanet", "Cabin", "Destination", "CryoSleep", "VIP"], drop_first=True)

In [161]:
dummies.shape
dummies.head()

Age  RoomService  FoodCourt  ShoppingMall     Spa  VRDeck  Transported  \
0  39.0          0.0        0.0           0.0     0.0     0.0        False   
1  24.0        109.0        9.0          25.0   549.0    44.0         True   
2  58.0         43.0     3576.0           0.0  6715.0    49.0        False   
3  33.0          0.0     1283.0         371.0  3329.0   193.0        False   
4  16.0        303.0       70.0         151.0   565.0     2.0         True   

   HomePlanet_Europa  HomePlanet_Mars  Cabin_B  Cabin_C  Cabin_D  Cabin_E  \
0               True            False     True    False    False    False   
1              False            False    False    False    False    False   
2               True            False    False    False    False    False   
3               True            False    False    False    False    False   
4              False            False    False    False    False    False   

   Cabin_F  Cabin_G  Cabin_T  Destination_PSO J318.5-22  \
0    False    False    False                      False   
1     True    False    False                      False   
2    False    False    False                      False   
3    False    False    False                      False   
4     True    False    False                      False   

   Destination_TRAPPIST-1e  CryoSleep_True  VIP_True  
0                     True           False     False  
1                     True           False     False  
2                     True           False      True  
3                     True           False     False  
4                     True           False     False

In [162]:
# Another way to do it: 
# categorical_cols = spaceship_cleaned.select_dtypes(include="object").columns
# spaceship_encoded = pd.get_dummies(spaceship_cleaned, columns=categorical_cols, drop_first=True)

In [163]:
# spaceship_cleaned.reset_index(drop=True, inplace=True)
# dummies.reset_index(drop=True, inplace=True)

**Perform Train Test Split**

In [164]:
features = dummies.drop(columns=["Transported"])
target = dummies["Transported"]

In [165]:
x_train, x_test, y_train, y_test = train_test_split(features, target, test_size=0.20, random_state=0)

**Model Selection**

In this exercise we will be using **KNN** as our predictive model.

In [166]:
normalizer = MinMaxScaler()

In [167]:
normalizer.fit(x_train)

MinMaxScaler()

In [168]:
x_train_norm = normalizer.transform(x_train)
x_test_norm = normalizer.transform(x_test)

In [169]:
x_train_norm

array([[4.05063291e-01, 0.00000000e+00, 0.00000000e+00, ...,
        1.00000000e+00, 1.00000000e+00, 0.00000000e+00],
       [5.06329114e-02, 0.00000000e+00, 0.00000000e+00, ...,
        1.00000000e+00, 1.00000000e+00, 0.00000000e+00],
       [3.79746835e-01, 0.00000000e+00, 7.91600979e-03, ...,
        1.00000000e+00, 0.00000000e+00, 0.00000000e+00],
       ...,
       [4.55696203e-01, 0.00000000e+00, 1.59527723e-01, ...,
        1.00000000e+00, 0.00000000e+00, 0.00000000e+00],
       [4.30379747e-01, 0.00000000e+00, 1.34169658e-04, ...,
        1.00000000e+00, 0.00000000e+00, 1.00000000e+00],
       [1.77215190e-01, 2.01612903e-04, 2.95508671e-02, ...,
        1.00000000e+00, 0.00000000e+00, 0.00000000e+00]])

In [170]:
x_train_norm = pd.DataFrame(x_train_norm, columns= x_train.columns)
x_train_norm

Age  RoomService  FoodCourt  ShoppingMall       Spa    VRDeck  \
0     0.405063     0.000000   0.000000      0.000000  0.000000  0.000000   
1     0.050633     0.000000   0.000000      0.000000  0.000000  0.000000   
2     0.379747     0.000000   0.007916      0.000000  0.051276  0.000000   
3     0.215190     0.001310   0.000000      0.046111  0.016378  0.000049   
4     0.329114     0.000000   0.000000      0.000000  0.000000  0.000000   
...        ...          ...        ...           ...       ...       ...   
5279  0.670886     0.000000   0.000000      0.000000  0.000000  0.000000   
5280  0.455696     0.000000   0.000000      0.000000  0.032355  0.000098   
5281  0.455696     0.000000   0.159528      0.000000  0.348893  0.004721   
5282  0.430380     0.000000   0.000134      0.000000  0.030569  0.087480   
5283  0.177215     0.000202   0.029551      0.000000  0.025794  0.010523   

      HomePlanet_Europa  HomePlanet_Mars  Cabin_B  Cabin_C  Cabin_D  Cabin_E  \
0                   0.0              1.0      0.0      0.0      0.0      1.0   
1                   0.0              0.0      0.0      0.0      0.0      0.0   
2                   0.0              0.0      0.0      0.0      0.0      0.0   
3                   0.0              1.0      0.0      0.0      0.0      1.0   
4                   0.0              0.0      0.0      0.0      0.0      0.0   
...                 ...              ...      ...      ...      ...      ...   
5279                0.0              1.0      0.0      0.0      0.0      1.0   
5280                0.0              0.0      0.0      0.0      0.0      0.0   
5281                1.0              0.0      0.0      0.0      1.0      0.0   
5282                1.0              0.0      0.0      0.0      1.0      0.0   
5283                1.0              0.0      0.0      0.0      1.0      0.0   

      Cabin_F  Cabin_G  Cabin_T  Destination_PSO J318.5-22  \
0         0.0      0.0      0.0                        0.0   
1         0.0      1.0      0.0                        0.0   
2         1.0      0.0      0.0                        0.0   
3         0.0      0.0      0.0                        0.0   
4         0.0      1.0      0.0                        0.0   
...       ...      ...      ...                        ...   
5279      0.0      0.0      0.0                        0.0   
5280      1.0      0.0      0.0                        0.0   
5281      0.0      0.0      0.0                        0.0   
5282      0.0      0.0      0.0                        0.0   
5283      0.0      0.0      0.0                        0.0   

      Destination_TRAPPIST-1e  CryoSleep_True  VIP_True  
0                         1.0             1.0       0.0  
1                         1.0             1.0       0.0  
2                         1.0             0.0       0.0  
3                         1.0             0.0       0.0  
4                         1.0             1.0       0.0  
...                       ...             ...       ...  
5279                      1.0             0.0       0.0  
5280                      0.0             0.0       0.0  
5281                      1.0             0.0       0.0  
5282                      1.0             0.0       1.0  
5283                      1.0             0.0       0.0  

[5284 rows x 19 columns]

In [171]:
x_test_norm = pd.DataFrame(x_test_norm, columns= x_test.columns)
x_test_norm.head()

Age  RoomService  FoodCourt  ShoppingMall      Spa  VRDeck  \
0  0.632911          0.0   0.000000           0.0  0.00000     0.0   
1  0.227848          0.0   0.000000           0.0  0.00000     0.0   
2  0.189873          0.0   0.000000           0.0  0.00000     0.0   
3  0.658228          0.0   0.000000           0.0  0.00000     0.0   
4  0.784810          0.0   0.054775           0.0  0.07774     0.0   

   HomePlanet_Europa  HomePlanet_Mars  Cabin_B  Cabin_C  Cabin_D  Cabin_E  \
0                0.0              1.0      0.0      0.0      0.0      0.0   
1                0.0              0.0      0.0      0.0      0.0      0.0   
2                0.0              0.0      0.0      0.0      0.0      0.0   
3                0.0              0.0      0.0      0.0      0.0      0.0   
4                1.0              0.0      1.0      0.0      0.0      0.0   

   Cabin_F  Cabin_G  Cabin_T  Destination_PSO J318.5-22  \
0      1.0      0.0      0.0                        0.0   
1      0.0      1.0      0.0                        1.0   
2      0.0      1.0      0.0                        0.0   
3      0.0      1.0      0.0                        1.0   
4      0.0      0.0      0.0                        0.0   

   Destination_TRAPPIST-1e  CryoSleep_True  VIP_True  
0                      0.0             1.0       0.0  
1                      0.0             1.0       0.0  
2                      0.0             1.0       0.0  
3                      0.0             1.0       0.0  
4                      1.0             0.0       1.0

In [172]:
knn = KNeighborsRegressor(n_neighbors=10)

In [173]:
knn.fit(x_train_norm, y_train)

KNeighborsRegressor(n_neighbors=10)

- Evaluate your model's performance. Comment it

In [174]:
knn.score(x_test_norm, y_test)

0.36844175491679265

In [175]:
pred = knn.predict(x_test_norm)
pred

array([1. , 0.9, 0.9, ..., 0.9, 0.7, 1. ])

In [176]:
from sklearn.metrics import mean_squared_error

MSE = mean_squared_error(y_test, pred)
RMSE = mean_squared_error(y_test, pred, squared=False)

print(MSE)
print(RMSE)

0.15788956127080184
0.3973531946150702


/opt/anaconda3/lib/python3.12/site-packages/sklearn/metrics/_regression.py:492: FutureWarning: 'squared' is deprecated in version 1.4 and will be removed in 1.6. To calculate the root mean squared error, use the function'root_mean_squared_error'.
  warnings.warn(


The Mean Squared Error (MSE) is 0.15789. This means that, on average, the squared difference between the actual and predicted values is about 0.15789. In general, lower MSE values indicate better model performance, with 0 being perfect.

The Root Mean Squared Error (RMSE) is approximately 0.3979. This is simply the square root of the MSE and gives a more interpretable measure — it tells us that, on average, the model's predictions are about 0.3979 units away from the actual values. Again, the lower this number, the better.

Overall, these error values suggest that the model is performing reasonably well, with relatively small prediction errors.